# Chapter 1 - Python Fundamentals for Chemical Engineers

## 1.1. Variables & Types, Math Operations, Casting

**Context:** Mini-scenarios from process/chemical engineering (mass/energy balance, mixing, Re number).

### Learning objectives
By the end of this notebook you will be able to:
- Declare variables with meaningful names and track **units**.
- Perform arithmetic with attention to **unit consistency** and operator precedence.
- Use `input()` and cast to `int`/`float` safely to avoid type errors.
- Build boolean expressions to check **specs, limits, and interlocks**.

> Tip: Add units in variable names or comments (e.g., `pressure_bar`, `T_C`) to avoid mistakes.

### Table of Contents

- [1.1. Variables & Types, Math Operations, Casting](#11-variables--types-math-operations-casting)
    - [1.1.1. Variables & Types (with units)](#111-variables--types-with-units)
        - [Naming Rules](#naming-rules)
        - [Data Types](#data-types)
        - [Type flexibility](#type-flexibility)
        - [Engineering Example](#engineering-example)
    - [1.1.2. Data Types](#112-data-types)
    - [1.1.3. Math Operations in Python and Chemical Engineering](#113-math-operations-in-python-and-chemical-engineering)
        - [Operator Precedence](#operator-precedence)
        - [Ideal Gas Law Example](#ideal-gas-law-example)
        - [Example: Mixing Tank (mass balance)](#example-mixing-tank-mass-balance)
        - [Example: Reynolds Number](#example--reynolds-number)
        - [Example — Filling Drums from a Tank](#example--filling-drums-from-a-tank)
        - [Example — Pressure Vessel Volume](#example--pressure-vessel-volume)
    - [1.1.4. Input & Type Casting](#114-input--type-casting)

### 1.1.1. Variables & Types (with units)

In Python, a **variable** stores a value.  
Always include the **units** in the name or as a comment — this is critical in engineering.

Examples:  
- `temperature_C` → temperature in °C  
- `pressure_bar` → pressure in bar  
- `Q_m3_h` → volumetric flow in m³/h  


In [2]:
# Reactor operating conditions
temperature_C = 85.0     # °C (reactor jacket outlet)
pressure_bar = 3.2       # bar (reactor headspace absolute)

# Unit conversions
temperature_K = temperature_C + 273.15     # K
pressure_Pa   = pressure_bar * 1e5         # Pa

print("T =", temperature_C, "°C  |", temperature_K, "K")
print("P =", pressure_bar, "bar |", pressure_Pa, "Pa")


T = 85.0 °C  | 358.15 K
P = 3.2 bar | 320000.0 Pa


#### Naming Rules

- Use **descriptive names**:
  - ✅ `viscosity_cP`, `density_kg_m3`
  - ❌ `x`, `data1` (unclear later)

- Rules:
  - Must start with a letter or `_`
  - No spaces (use `_`)
  - Case-sensitive (`T` ≠ `t`)

### 1.1.2. Data Types

Python has several basic types useful in engineering:

- **int** → integers (e.g., `5`)
- **float** → decimals (e.g., `3.14`)
- **str** → text/strings (e.g., `"Reactor A"`)
- **bool** → logical values `True` or `False`

Check a variable’s type with `type()`:

In [3]:
print(type(997))          # int
print(type(0.00089))      # float
print(type("kg/m^3"))     # str
print(type(True))         # bool

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


#### Type flexibility

Python can combine types in calculations:

- int × float → float  

In [4]:
flow = 10        # int
time = 3.5       # float
volume = flow * time   # result is float

⚠️ Be careful: "100" is a string, not a number!

"100" + 50 → ❌ error

int("100") + 50 → ✅ 150

#### Engineering Example

Calculate the mass of oil in a storage tank:

- Density = 850 kg/m³  
- Volume = 2.5 m³

In [5]:
density_kg_m3 = 850.0
V_m3 = 2.5
mass_kg = density_kg_m3 * V_m3
print("Mass =", mass_kg, "kg")

Mass = 2125.0 kg


### 1.1.3. Math Operations in Python and Chemical Engineering

Mathematics is the language of engineering. Python uses symbols that are similar to what we
write on paper, but there are some important details:

| Symbol | Operation                | Example (Python) | Example (Engineering) |
|--------|---------------------------|------------------|------------------------|
| `+`    | Addition                  | `2 + 3` → 5      | Add two flowrates: `Q_total = Q1 + Q2` |
| `-`    | Subtraction               | `5 - 2` → 3      | Pressure drop: `ΔP = Pin - Pout` |
| `*`    | Multiplication            | `4 * 2` → 8      | Heat duty: `Q = m * cp * ΔT` |
| `/`    | Division                  | `10 / 4` → 2.5   | Velocity: `u = Q / A` |
| `**`   | Exponent (power)          | `3 ** 2` → 9     | Area of circle: `A = π * r**2` |
| `%`    | Modulo (remainder)        | `10 % 3` → 1     | Rare in engineering, but useful in cycles/schedules |
| `//`   | Floor division (integer)  | `10 // 3` → 3    | Batch count: number of full 200 L drums from 650 L |

#### Operator Precedence

Like in algebra, Python respects the order of operations:
1. Parentheses `( )`
2. Exponentiation `**`
3. Multiplication/division `*`, `/`, `//`, `%`
4. Addition/subtraction `+`, `-`

For example:  


In [6]:
result1 = 100 / 2 * 5      # left to right: (100/2)=50 → (50*5)=250
result2 = 100 / (2 * 5)    # parentheses: 100/10 = 10
print("Result 1 =", result1)
print("Result 2 =", result2)

Result 1 = 250.0
Result 2 = 10.0


This is important in engineering calculations. For example, when applying correlations like
the Reynolds number or Nusselt correlations, forgetting parentheses can give results off by
orders of magnitude.

### Example - Ideal Gas Law

**Scenario:** You have a reactor vessel with:
- Pressure = 2.5 bar (absolute)
- Volume = 1.2 m³
- Temperature = 30 °C

**Goal:** Estimate how many moles of gas are inside, using the Ideal Gas Law:

$
PV = nRT \quad \Rightarrow \quad n = \frac{PV}{RT}
$

Where:
- $P$ is the pressure (bar)
- $V$ is the volume (L)
- $R$ = 0.08314, $L.bar.mol^{-1}.K^{-1}$
- $T$ is the temperature (K)

Notice we had to **convert the volume from m³ to L** and **temperature to K**.  
This reinforces why *unit consistency* is essential in engineering.

In [7]:
P_bar = 2.5                # bar
V_m3  = 1.2                # m³
T_C   = 30.0               # °C
R = 0.08314                # L·bar/(mol·K)

T_K = T_C + 273.15
V_L = V_m3 * 1000.0        # convert to liters

n_mol = (P_bar * V_L) / (R * T_K)
print(f"Estimated moles of gas: {n_mol:.2f} mol")

Estimated moles of gas: 119.03 mol


### Example - Mixing Tank (mass balance)

**Scenario:**  
Imagine a mixing tank in a pilot plant where two liquid streams enter and mix completely before leaving.  
- Stream 1: low concentration of ethanol in water  
- Stream 2: higher concentration of ethanol in water  
- The tank is perfectly mixed and operates at steady state (so accumulation = 0).  

**Mass balance principle:**  

At steady state:
$
\text{Input} = \text{Output}
$

For a solute (e.g., ethanol), this becomes:
$
Q_1 C_1 + Q_2 C_2 = (Q_1 + Q_2) C_{out}
$

Rearranging gives the outlet concentration:
$
C_{out} = \frac{Q_1C_1 + Q_2C_2}{Q_1 + Q_2}
$

Where:  
- $ Q_i $ = volumetric flow rate of stream *i* (e.g., L/min)  
- $ C_i $ = solute concentration in stream *i* (g/L)  
- $ C_{out} $ = concentration in outlet stream (g/L)  

**Interpretation:**  
- If **Stream 1 dominates the flow**, the outlet concentration will be close to $C_1$.  
- If both streams are equal, the outlet will be the **simple average** of $C_1$ and $C_2$.  
- This formula is essentially a **weighted average**, where the "weights" are the flowrates.  

💡 This is one of the most common calculations in chemical engineering, and it appears in mixing, blending, and even in the calculation of average properties in reactors and separation units.

In [8]:
Q1_L_min, C1_g_L = 8.0, 0.12
Q2_L_min, C2_g_L = 5.0, 0.35

Q_total = Q1_L_min + Q2_L_min
solute_in_g_per_min = Q1_L_min*C1_g_L + Q2_L_min*C2_g_L
C_out = solute_in_g_per_min / Q_total

print(f"Q_total = {Q_total:.1f} L/min")
print(f"Solute in = {solute_in_g_per_min:.2f} g/min")
print(f"C_out = {C_out:.4f} g/L")

Q_total = 13.0 L/min
Solute in = 2.71 g/min
C_out = 0.2085 g/L


### Example - Reynolds Number

**Scenario:**  
Suppose you are designing a cooling water line. The pipe has an internal diameter of 1 inch (0.0254 m), and water flows with an average velocity of 1.2 m/s at 25 °C.  
We want to know if the flow is laminar or turbulent, since this affects **pressure drop** and **heat transfer rates**.

---

**Formula:**

$Re = \dfrac{\rho \, u \, D}{\mu}$

Where:  
- $\rho$ = density of the fluid (kg/m³)  
- $u$ = average velocity (m/s)  
- $D$ = pipe diameter (m)  
- $\mu$ = viscosity of the fluid (Pa·s)

---

**Given values (water at ~25 °C):**

- $\rho$ = 997 $kg/m³$  
- $u$ = 1.2  $m/s$  
- $D$ = 0.0254  $m$  
- $\mu$ = 0.00089 $Pa.s$  

**Interpretation:**

- $Re < 2100$ → Laminar flow  
- $2100 < Re < 4000$ → Transition regime  
- $Re > 4000$ → Turbulent flow  

In [9]:
rho = 997.0       # kg/m³ (water at 25 °C)
u   = 1.2         # m/s
D   = 0.0254      # m (1 inch pipe)
mu  = 0.00089     # Pa·s

Re = (rho * u * D) / mu

print(f"Reynolds number = {Re:,.0f}")


Reynolds number = 34,144


### Example - Filling Drums from a Tank

You need to package 650 L of solvent into 200 L drums.

- Floor division (`//`) tells us how many **full drums** we can fill.  
- Modulo (`%`) tells us how much solvent is **left over**.


In [10]:
total_volume_L = 650
drum_capacity_L = 200

full_drums = total_volume_L // drum_capacity_L   # floor division
leftover = total_volume_L % drum_capacity_L      # remainder

print("Full drums:", full_drums)
print("Leftover solvent:", leftover, "L")


Full drums: 3
Leftover solvent: 50 L


### Example - Pressure Vessel Volume

Suppose you have a **spherical vessel** of internal radius $r = 1.2$ m.  
The volume is given by:

$
V = \frac{4}{3} \pi r^3
$

Notice the cube: $r^3$ is written in Python as `r**3`.


In [11]:
# Using pi = 3.14 (approximate value)
pi = 3.14
r_m = 1.2
V_m3_approx = (4/3) * pi * (r_m**3)
print(f"Vessel volume (using pi=3.14) = {V_m3_approx:.3f} m³")

# Using math.pi and explaining why it's better
import math
V_m3_math = (4/3) * math.pi * (r_m**3)
print(f"Vessel volume (using math.pi) = {V_m3_math:.3f} m³")
print("Using external packages like 'math' ensures higher precision and reliability, especially for scientific constants like pi.")


Vessel volume (using pi=3.14) = 7.235 m³
Vessel volume (using math.pi) = 7.238 m³
Using external packages like 'math' ensures higher precision and reliability, especially for scientific constants like pi.


Other common uses of exponents in chemical engineering:

- **Arrhenius equation**: $k = A \, e^{-E_a/(RT)}$ (exponential)  
- **Correlations**: friction factor, Nusselt number, etc. often use $Re^n$ or $Pr^m$  
- **Polytropic processes**: $PV^n = \text{constant}$ involves exponents  

The exponent operator is therefore fundamental in modeling and correlations.

## 1.1.4. Input & Type Casting

So far, all numbers in our calculations were *hard-coded* inside the program (we typed them directly into variables).  
But in real engineering work, values often come from **outside sources**:
- An operator typing a flowrate into an interface
- A sensor measurement saved as text in a file
- A colleague entering experimental data manually

In Python, the function `input()` is how we collect values from the user.  

⚠️ Important: everything that comes from `input()` is treated as **text** (a *string*), even if it looks like a number.  
This means we need to **convert** (or *cast*) the text into the right type before we can use it in calculations.

---

### Why this matters in engineering

Imagine reading `"25"` from a file:  
- If Python thinks it is text, it cannot be multiplied by density to get mass.  
- Once we cast it into a number (`int` or `float`), it becomes usable in equations.

This is why **type casting** is a basic but essential skill for engineers who use Python to process data.

### Example - Tank Mass Calculation

Suppose an operator measures the volume of liquid in a storage tank.  
We want to compute the **mass** using density.

$
m = \rho \, V
$

Where:  
- $\rho$ = density (kg/m³)  
- $V$ = volume (m³)  

Since the operator types the values, we must **cast them to floats** before multiplying.



In [12]:
density = float(input("Enter density (kg/m³): "))
volume = float(input("Enter volume (m³): "))

mass = density * volume
print(f"Density = {density:.2f} kg/m³")
print(f"Volume = {volume:.2f} m³")
print(f"Mass of liquid = {mass:.2f} kg")


Mass of liquid = 2000.00 kg


### Example - Energy Balance (Heating Water)

We want to compute the **heat duty** for heating a stream of water.

$
Q = \dot{m} \, c_p \, \Delta T
$

Where:  
- $\dot{m}$ = mass flowrate (kg/s)  
- $c_p = 4.18 \, \text{kJ/(kg.K)}$  
- $\Delta T = T_{out} - T_{in}$  

The user provides the flowrate and temperatures.


In [13]:
m_dot = float(input("Enter mass flowrate (kg/s): "))
T_in = float(input("Enter inlet temperature (°C): "))
T_out = float(input("Enter outlet temperature (°C): "))
cp = 4.18   # kJ/(kg·K)

Q_kW = m_dot * cp * (T_out - T_in)   # result in kW

print(f"Mass flowrate = {m_dot:.2f} kg/s")
print(f"Inlet temperature = {T_in:.2f} °C")
print(f"Outlet temperature = {T_out:.2f} °C")
print(f"Specific heat capacity = {cp:.2f} kJ/(kg·K)")
print(f"Heat duty = {Q_kW:.2f} kW")


Mass flowrate = 12.00 kg/s
Inlet temperature = 23.00 °C
Outlet temperature = 105.00 °C
Specific heat capacity = 4.18 kJ/(kg·K)
Heat duty = 4113.12 kW


### Example - Pipe Velocity

For a given volumetric flowrate $Q$ and pipe diameter $D$, the velocity is:

$
u = \frac{Q}{A}, \quad A = \frac{\pi D^2}{4}
$

We ask the user to type $Q$ (in m³/h) and $D$ (in m), convert units, and compute velocity.


In [ ]:
import math

Q_m3_h = float(input("Enter flowrate Q (m³/h): "))
D_m = float(input("Enter pipe diameter D (m): "))

Q_m3_s = Q_m3_h / 3600.0
A = math.pi * (D_m**2) / 4.0
u = Q_m3_s / A

print(f"Velocity = {u:.3f} m/s")
print(f"Cross-sectional area = {A:.3f} m²")
print(f"Flowrate = {Q_m3_s:.3f} m³/s")
print(f"Pipe diameter = {D_m:.3f} m")
print(f"Volumetric flowrate = {Q_m3_h:.3f} m³/h")


## 1.1.5. Booleans & Comparisons

In chemical and process engineering, we constantly need to **make decisions based on conditions**:
- Is the product inside specifications?  
- Is the reactor safe to operate?  
- Should an alarm be triggered?  
- Is a pump running while the valve is closed?

These yes/no questions are naturally represented in Python by the **boolean type**, which has only two values:
- `True`  
- `False`

---

### Why this matters in engineering

- **Quality control:** release a batch only if density and viscosity are within range.  
- **Process safety:** shut down equipment if temperature or pressure exceed limits.  
- **Automation:** trigger alarms if a tank level is too high or too low.  

Booleans are the bridge between **numerical calculations** (flows, pressures, temperatures) and **logical decisions** (alarms, shutdowns, batch release).  

---

### Operators

**Comparison operators** let us test conditions:
- `>`  greater than  
- `<`  less than  
- `>=` greater or equal  
- `<=` less or equal  
- `==` equal  
- `!=` not equal  

**Logical operators** combine conditions:
- `and` → both conditions must be true  
- `or` → at least one condition must be true  
- `not` → inverts the condition  

Together, these let us encode the same kinds of rules we use in **plant operation manuals, interlock systems, and QA checklists**.

### Example - Distillation Column Overpressure (`>`)

A distillation column has a maximum operating pressure of 6 bar.  
If the actual pressure is **greater than** 6, a high-pressure alarm must be raised.


In [19]:
# True case: pressure exceeds the maximum
pressure_bar = 6.5
max_pressure_bar = 6.0
high_pressure_alarm = pressure_bar > max_pressure_bar
print("High-pressure alarm triggered? (True case)", high_pressure_alarm)

# False case: pressure does not exceed the maximum
pressure_bar = 5.8
high_pressure_alarm = pressure_bar > max_pressure_bar
print("High-pressure alarm triggered? (False case)", high_pressure_alarm)

High-pressure alarm triggered? (True case) True
High-pressure alarm triggered? (False case) False


### Example - Cooling Tower Low Flow (`<`)

A cooling tower must maintain water flow above 100 m³/h.  
If flow is **less than** 100, efficiency drops and a low-flow alarm is triggered.


In [15]:
# True case: flow is less than minimum, alarm should trigger
flow_m3_h = 85
min_flow = 100

low_flow_alarm = flow_m3_h < min_flow
print("Low-flow alarm triggered? (True case)", low_flow_alarm)

# False case: flow is above minimum, alarm should not trigger
flow_m3_h = 120
low_flow_alarm = flow_m3_h < min_flow
print("Low-flow alarm triggered? (False case)", low_flow_alarm)


Low-flow alarm triggered? True


### Example - Tank Full Detection (`>=`)

A storage tank is considered “full” if its level reaches **at least 95%** of capacity.  
If the level is **greater than or equal to** 95%, the filling pump must be stopped.


In [20]:
# True case: tank level is at the threshold, pump should stop
tank_level_percent = 95
full_threshold = 95

tank_full = tank_level_percent >= full_threshold
print("Stop filling pump? (True case)", tank_full)

# False case: tank level is below the threshold, pump should not stop
tank_level_percent = 90
tank_full = tank_level_percent >= full_threshold
print("Stop filling pump? (False case)", tank_full)

Stop filling pump? (True case) True
Stop filling pump? (False case) False


### Example - Heat Exchanger Low Temperature (`<=`)

A heat exchanger should never cool the stream below 40 °C to avoid condensation.  
If the outlet temperature is **less than or equal to** 40, a warning must be issued.


In [21]:
# True case: outlet temperature is less than or equal to minimum, warning should trigger
T_out = 39.5
T_min = 40.0

too_cold = T_out <= T_min
print("Temperature too low warning? (True case)", too_cold)

# False case: outlet temperature is above minimum, warning should not trigger
T_out = 42.0
too_cold = T_out <= T_min
print("Temperature too low warning? (False case)", too_cold)


Temperature too low warning? (True case) True
Temperature too low warning? (False case) False


### Example - Batch Neutralization pH Check (`==`)

In a neutralization step, the batch must reach exactly pH = 7.0 before moving on.  
If the pH is **equal to** 7.0, the batch can proceed.


In [22]:
# True case: pH is exactly at the required value
pH = 7.0
required_pH = 7.0
ready_to_proceed = pH == required_pH
print("pH exactly adjusted? (True case)", ready_to_proceed)

# False case: pH is not at the required value
pH = 6.8
ready_to_proceed = pH == required_pH
print("pH exactly adjusted? (False case)", ready_to_proceed)


pH exactly adjusted? (True case) True
pH exactly adjusted? (False case) False


### Example - Sensor Disagreement (`!=`)

Two redundant sensors measure reactor pressure.  
If their readings are **not equal**, calibration or maintenance is required.


In [23]:
# True case: sensors disagree, calibration needed
sensor1 = 5.2
sensor2 = 5.6
calibration_needed = sensor1 != sensor2
print("Calibration required? (True case)", calibration_needed)

# False case: sensors agree, calibration not needed
sensor1 = 5.2
sensor2 = 5.2
calibration_needed = sensor1 != sensor2
print("Calibration required? (False case)", calibration_needed)


Calibration required? (True case) True
Calibration required? (False case) False


### Example - Reactor Safety Interlock (`or`)

A reactor trips if **either** pressure exceeds its limit **or** temperature exceeds its limit.  
Use `or` because **any single** unsafe condition should trigger the shutdown.


In [25]:
# Case 1: Both pressure and temperature exceed their limits (shutdown)
pressure_bar = 5.3
temperature_C = 121.0
max_pressure_bar = 5.0
max_temperature_C = 120.0

trip = (pressure_bar > max_pressure_bar) or (temperature_C > max_temperature_C)
print("Emergency shutdown triggered? (Both exceed)", trip)

# Case 2: Only pressure exceeds the limit (shutdown)
pressure_bar = 5.3
temperature_C = 115.0
trip = (pressure_bar > max_pressure_bar) or (temperature_C > max_temperature_C)
print("Emergency shutdown triggered? (Pressure only)", trip)

# Case 3: Only temperature exceeds the limit (shutdown)
pressure_bar = 4.8
temperature_C = 121.0
trip = (pressure_bar > max_pressure_bar) or (temperature_C > max_temperature_C)
print("Emergency shutdown triggered? (Temperature only)", trip)

# Case 4: Both pressure and temperature are within limits (no shutdown)
pressure_bar = 4.8
temperature_C = 115.0
trip = (pressure_bar > max_pressure_bar) or (temperature_C > max_temperature_C)
print("Emergency shutdown triggered? (Neither)", trip)


Emergency shutdown triggered? (Both exceed) True
Emergency shutdown triggered? (Pressure only) True
Emergency shutdown triggered? (Temperature only) True
Emergency shutdown triggered? (Neither) False


### Example - Batch Release QC (`and`)

A product is releasable **only if** density is within 980–1020 kg/m³ **and** viscosity ≤ 12 cP.  
Use `and` because **both** quality conditions must be satisfied.


In [26]:
density = 1008.0     # kg/m³
viscosity = 11.6     # cP

# Case 1: Both in spec (should release)
in_density_range = (980.0 <= density <= 1020.0)
in_viscosity_spec = (viscosity <= 12.0)
release_batch = in_density_range and in_viscosity_spec
print("Case 1 - Both in spec: Release batch?", release_batch)

# Case 2: Density in spec, viscosity out of spec (should not release)
viscosity = 13.0
in_viscosity_spec = (viscosity <= 12.0)
release_batch = in_density_range and in_viscosity_spec
print("Case 2 - Density ok, viscosity high: Release batch?", release_batch)

# Case 3: Density out of spec, viscosity in spec (should not release)
density = 970.0
viscosity = 11.6
in_density_range = (980.0 <= density <= 1020.0)
in_viscosity_spec = (viscosity <= 12.0)
release_batch = in_density_range and in_viscosity_spec
print("Case 3 - Density low, viscosity ok: Release batch?", release_batch)

# Case 4: Both out of spec (should not release)
density = 970.0
viscosity = 13.0
in_density_range = (980.0 <= density <= 1020.0)
in_viscosity_spec = (viscosity <= 12.0)
release_batch = in_density_range and in_viscosity_spec
print("Case 4 - Both out of spec: Release batch?", release_batch)


Case 1 - Both in spec: Release batch? True
Case 2 - Density ok, viscosity high: Release batch? False
Case 3 - Density low, viscosity ok: Release batch? False
Case 4 - Both out of spec: Release batch? False


### Example - Pump Operation Interlock (`and` + `not`)

To **start** a transfer pump, two conditions must hold:  
1) The suction valve is open.  
2) The discharge valve is open.  

If the pump is running **and** the discharge valve is **not** open, raise an overpressure risk.


In [27]:
pump_running   = True
suction_open   = True
discharge_open = True

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 1: Pump running, suction open, discharge open")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = True
suction_open   = True
discharge_open = False

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 2: Pump running, suction open, discharge closed")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = True
suction_open   = False
discharge_open = True

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 3: Pump running, suction closed, discharge open")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = True
suction_open   = False
discharge_open = False

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 4: Pump running, suction closed, discharge closed")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = False
suction_open   = True
discharge_open = True

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 5: Pump stopped, suction open, discharge open")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = False
suction_open   = True
discharge_open = False

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 6: Pump stopped, suction open, discharge closed")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = False
suction_open   = False
discharge_open = True

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 7: Pump stopped, suction closed, discharge open")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)
print()

pump_running   = False
suction_open   = False
discharge_open = False

can_start = suction_open and discharge_open
overpressure_risk = pump_running and (not discharge_open)
print("Case 8: Pump stopped, suction closed, discharge closed")
print("Can start pump?", can_start)
print("Overpressure risk?", overpressure_risk)


Case 1: Pump running, suction open, discharge open
Can start pump? True
Overpressure risk? False

Case 2: Pump running, suction open, discharge closed
Can start pump? False
Overpressure risk? True

Case 3: Pump running, suction closed, discharge open
Can start pump? False
Overpressure risk? False

Case 4: Pump running, suction closed, discharge closed
Can start pump? False
Overpressure risk? True

Case 5: Pump stopped, suction open, discharge open
Can start pump? True
Overpressure risk? False

Case 6: Pump stopped, suction open, discharge closed
Can start pump? False
Overpressure risk? False

Case 7: Pump stopped, suction closed, discharge open
Can start pump? False
Overpressure risk? False

Case 8: Pump stopped, suction closed, discharge closed
Can start pump? False
Overpressure risk? False


### Example - Analyzer Validation (`not`)

A gas analyzer can be used only when it is **validated**.  
Use `not` to block operation when validation is missing.


In [28]:
# Case 1: Analyzer validated (should NOT block operation)
analyzer_validated = True
block_operation = not analyzer_validated
print("Case 1: Analyzer validated")
print("Block operation due to invalid analyzer?", block_operation)
print()

# Case 2: Analyzer NOT validated (should block operation)
analyzer_validated = False
block_operation = not analyzer_validated
print("Case 2: Analyzer NOT validated")
print("Block operation due to invalid analyzer?", block_operation)


Case 1: Analyzer validated
Block operation due to invalid analyzer? False

Case 2: Analyzer NOT validated
Block operation due to invalid analyzer? True


## 1.1.6. Chapter Summary

This chapter introduced essential Python concepts for chemical and process engineering applications:

### Variables & Types
- Use descriptive variable names and always track **units** (e.g., `pressure_bar`, `T_C`).
- Python supports `int`, `float`, `str`, and `bool` types. Use `type()` to check a variable’s type.
- Be careful with type flexibility and conversions (e.g., casting strings to numbers for calculations).

### Math Operations
- Python uses familiar math symbols: `+`, `-`, `*`, `/`, `**`, `%`, `//`.
- Operator precedence follows algebraic rules (parentheses, exponents, multiplication/division, addition/subtraction).
- Always ensure **unit consistency** in calculations.
- Common engineering examples: mass/energy balances, mixing, Reynolds number, vessel volumes.

### Input & Type Casting
- Use `input()` to collect user data (always as a string).
- Cast input to `int` or `float` before calculations.
- Essential for real-world engineering where data comes from operators, sensors, or files.

### Booleans & Comparisons
- Boolean logic (`True`/`False`) is used for decision-making in engineering (alarms, interlocks, QA).
- **Comparison operators:** `>`, `<`, `>=`, `<=`, `==`, `!=`.
- **Logical operators:** `and`, `or`, `not`.
- Combine conditions to encode process rules, safety interlocks, and quality checks.

### Engineering Scenarios
- Each concept was illustrated with practical examples: pressure alarms, flow checks, tank levels, batch release, pump interlocks, analyzer validation, and more.
- Emphasis on translating engineering logic and calculations into clear, reliable Python code.

---

**Key Takeaway:**

Mastering these Python fundamentals enables you to automate calculations, implement process logic, and build robust tools for chemical engineering tasks.

## **Exercises**

### **1. Variables & Types**

1. Define variables for a distillation column: top temperature (in °C), bottom temperature (in °C), and reflux ratio (dimensionless). Print their values and types.
2. Store the density of water (1000 kg/m³) and ethanol (789 kg/m³) in variables. Calculate the mass of 2.5 m³ of each.
3. Assign a string `"Reactor A"` to a variable, and a boolean `True` to indicate it is operating. Print both.
4. Convert the string `"120"` into an integer and multiply it by 5 to simulate converting operator input into a usable flowrate.
5. Define `int`, `float`, `str`, and `bool` variables that describe a pump (e.g., number of stages, efficiency, tag name, running state). Print all types with `type()`.
6. Assign `pressure_bar = 5` and `temperature_C = 120.5`. Convert them to Pa and K respectively and print results with units.
7. A tank level indicator shows `"75"` (string). Convert it to `int` and compute the fraction of tank filled if total capacity = 100.

---

### **2. Math Operations**

8. Compute the total mass in kg of a liquid stream with flowrate 12 kg/s running for 5 minutes.
9. A heat exchanger raises water temperature from 25 °C to 85 °C at 1.5 kg/s. Calculate heat duty with $Q = \dot m c_p \Delta T$ $(c_p = 4.18\$ kJ/kg.K)$.
10. A reactor vessel has radius 1.2 m. Calculate volume using $V = \frac{4}{3} \pi r^3$.
11. Two streams mix: 6 L/min at 0.10 g/L and 4 L/min at 0.40 g/L. Calculate outlet concentration.
12. For flow $u=1.0$ m/s, pipe $D=0.05$ m, $\rho=1000$ kg/m³, $\mu=0.001$ Pa·s, calculate Reynolds number. Classify regime.
13. Package 725 L of product into 200 L drums. Use `//` to compute full drums, and `%` to compute leftover volume.
14. A process takes samples every 8 min for 120 min. Use floor division and modulo to find number of samples and leftover minutes.
15. Show the difference between `100/2*5` and `100/(2*5)`. Discuss why operator precedence matters in engineering equations.

---

### **3. Input & Type Casting**

16. Ask the user for density (kg/m³) and volume (m³), cast to float, and compute mass.
17. Request mass flowrate (kg/s) and inlet/outlet temperatures (°C), then compute heat duty.
18. Ask the user for flow (m³/h) and pipe diameter (m). Compute velocity in m/s.
19. Request pH from user. Print if water is potable (6.5–8.5).
20. Ask user for tank level (%) as string, cast to integer, and check if above 90%.
21. Input ethanol mass fraction (string), cast to float, then compute water fraction as `1 - x_ethanol`.
22. Ask user for two stream flows (L/min) and concentrations (g/L). Compute outlet concentration.

---

### **4. Booleans & Comparisons**

23. A batch is acceptable if viscosity is between 5–9 cP. Test with viscosity = 8.2.
24. Reactor trips if pressure > 5 bar **or** temperature > 120 °C. Test with P=4.9, T=125.
25. A tank is considered full if level ≥ 95%. Test with 97%.
26. A cooler is unsafe if outlet ≤ 40 °C. Test with T\_out=38.5.
27. A batch step requires exactly pH = 7.0. Check for pH=7.2 and pH=7.0.
28. Two thermocouples read 150 °C and 154 °C. If values are not equal, calibration is needed.
29. A storage tank is critical if level < 10% **or** > 90%. Test with 5% and 60%.
30. A pump can run only if suction and discharge valves are both open. Test with suction open=True, discharge open=False.

---

## Challenges

### 1. Refinery Gasoline Blending

A refinery wants to meet an octane specification by blending two gasoline streams volumetrically. Stream A has octane 75, Stream B has octane 86; the desired blend octane is 83. Find the required volume fraction of Stream B using the relation:

$$ON_{blend} = xON_{B} + (1-x)ON_{A}$$

Solve for $x$. (Inspired by linear blending models described in fuel refinery contexts.) - Source: [chegg](https://www.chegg.com/homework-help/questions-and-answers/problem-2-liquid-water-fed-boiler-t1-c-p1-bars-absolute-converted-saturated-steam-p1-bars--q33495986?utm_source=chatgpt.com)


### 2. Pump Power Calculation (SI Units)

Design a pump transporting water (density = 1000 kg/m³) at a flow rate of 1 m³/h against a head of 10 m. Compute the hydraulic power:

$$P_{hydraulic} = \frac{\rho g Q H}{3.6\times10^{6}}\text{ }kW$$

Then compute shaft power given efficiency 80%. Sources: [engineeringtoolbox](https://www.engineeringtoolbox.com/pumps-power-d_505.html?utm_source=chatgpt.com), [wikipedia](https://en.wikipedia.org/wiki/Centrifugal_pump?utm_source=chatgpt.com)

### 3. Steam Heating Duty

Steam is widely used in chemical plants to heat process streams because when it condenses, it releases a large amount of latent heat (~2250 kJ/kg). To size steam utilities, we first calculate the required heat duty for the process stream, then divide by steam latent heat to estimate steam consumption.

A process requires heating water from 25 °C to 80 °C at a flow rate of 10 kg/s.
The specific heat of water is $c_p = 4.18 ,\text{kJ/(kg.K)}$.

Calculate the necessary mass flow rate of steam to supply the demanded duty. Express the results in $kg.s^{-1}$ e $kg.h^{-1}$.

Sources: [Engineering Toolbox](https://www.engineeringtoolbox.com/steam-heating-process-d_437.html)

### 4. Milk Pasteurization Duty (HTST)

Milk enters a pasteurizer at 4 °C and must be heated to 72 °C. Flow rate is 500 L/min, density ≈ 1 kg/L, cp ≈ 4.0 kJ/kg·K. Compute the heat duty.
Include a note that HTST standard is 72 °C for 15 s. 

Sources: [Thermtest](https://thermtest.com/applications-of-heat-exchangers-in-the-dairy-industry?utm_source=chatgpt.com), [ScienceDirect](https://www.sciencedirect.com/science/article/abs/pii/S0956713503001841?utm_source=chatgpt.com)

### 5. Pasteurization Energy Recovery

In a pasteurizer with regeneration sections that recover 82% of heat, the specific heat consumption for milk processing is reduced to ~51.7 kJ/kg. Compute the actual heating duty for a flow of 1000 kg/hr of milk. 

Sources: [ResearchGate](https://www.researchgate.net/publication/269474572_Heat_Consumption_and_Quality_of_Milk_Pasteurization?utm_source=chatgpt.com)

### 6. Water Treatment Chlorine Dosing

A wastewater plant treats 100 L/s of water with chlorine at 2 mg/L. Calculate the daily amount of chlorine required in kg per day using 100 L/s . 2 mg/L . 86,400 s/day

### 7. Flammability Safety Check (Tank Vapor)

A storage tank headspace contains a flammable component at 3.0 % v/v. The LFL is 1.2 %. Write a boolean expression to check whether the mixture is in the flammable range

Source: [Wikipedia](https://en.wikipedia.org/wiki/Octane_rating?utm_source=chatgpt.com)

### 8. Batch Reactor Conversion (Pharma)

A batch reactor converts 80% of reactant A into product P with 1:1 stoichiometry. If the initial mass of A is 500 kg, compute the product mass. Consider the same molar basis for A and P.

### 9. Distillation Recovery Balance

A distillation feed of 1000 kg/h contains 40% component A. The distillate purity is 95% A and over 90% of A is recovered. Compute distillate (D) and bottoms (B) via mass balance. 

### 10. Pump System Power vs. Motor Rating

Extend Exercise 2: If the calculated required electrical power exceeds motor rating of 20 kW, trigger a boolean indicating "upgrade required".